Kernel/FileSystem/FATFileSystem.cpp

```diff
@@ -0,0 +1,361 @@
+            filename.append(lfn_entry.characters2[4]);
+            filename.append(lfn_entry.characters2[5]);
+            filename.append(lfn_entry.characters3[0]);
+            filename.append(lfn_entry.characters3[1]);
+        }
+
+        m_filename = MUST(KString::try_create(byte_terminated_string(filename.string_view().bytes().data(), filename.length(), 0xFF)));
+    }
+
+    dbgln_if(FAT_VERY_DEBUG, "FATFS: Computed filename {} for inode {}", m_filename, index());
+}
+
+time_t FATInode::fat_date_time(u16 date, u16 time) const
+{
+    time_t year = (date >> 9) + (1980 - 1970);
+    time_t month = (date >> 5) & 0b1111;
+    time_t day = date & 0b11111;
+
+    time_t hour = time >> 11;
+    time_t minute = (time >> 5) & 0b1111111;
+    time_t second = (time & 0b11111) << 1;
+
+    return second + minute * 60 + hour * 3600
+        + day + (24 * 3600) + month * (30 * 24 * 3600) + year * (365 * 24 * 3600);
+}
+
+// TODO: This might be worth putting as a part of the StringView class
+StringView FATInode::byte_terminated_string(u8 const* string, size_t max_length, u8 byte) const
+{
+    size_t length = max_length;
+
+    // HACK: size_t is unsigned, so we can't check if it's smaller than 0,
+    //       but when we go below zero the value turns really big
+    while (length <= max_length) {
+        if (string[length - 1] != byte) {
+            return StringView(string, length);
+        }
+        length -= 1;
+    }
+    return StringView(string, max_length);
+}
+
+ErrorOr<size_t> FATInode::read_bytes_locked(off_t offset, size_t size, UserOrKernelBuffer& buffer, OpenFileDescription*) const
+{
+    dbgln_if(FAT_DEBUG, "FATFS: Reading inode {}: ize: {} offset: {}", identifier().index(), size, offset);
+
+    // FIXME: Read only the neaded blocks instead of the whole file
+    auto blocks = TRY(const_cast<FATInode&>(*this).read_block_list());
+    TRY(buffer.write(blocks->data() + offset, min(size, m_block_list.size() * fs().m_logical_block_size - offset)));
+
+    return min(size, m_block_list.size() * fs().m_logical_block_size - offset);
+}
+
+InodeMetadata FATInode::metadata() const
+{
+    return m_metadata;
+}
+
+ErrorOr<void> FATInode::traverse_as_directory(Function<ErrorOr<void>(FileSystem::DirectoryEntryView const&)> callback) const
+{
+    MutexLocker locker(m_inode_lock);
+
+    VERIFY(m_entry.attributes & FATAttributes::Directory);
+
+    Vector<RefPtr<FATInode>> children = TRY(const_cast<FATInode&>(*this).traverse());
+
+    for (auto const& inode : children) {
+        TRY(callback({ inode->m_filename->view(), inode->identifier(), inode->m_entry.attributes }));
+    }
+
+    return {};
+}
+
+ErrorOr<NonnullLockRefPtr<Inode>> FATInode::lookup(StringView name)
+{
+    MutexLocker locker(m_inode_lock);
+
+    VERIFY(m_entry.attributes & FATAttributes::Directory);
+
+    Vector<RefPtr<FATInode>> children = TRY(traverse());
+
+    for (auto& inode : children) {
+        // FIXME: For some reason using == doesn't work
```

Comment:
What are you trying to use `operator==` on? The `inode` specifically? It doesn't look like you have `operator==()` implemented in your `FATInode` class, did you perhaps mean to implement that?

Reply:
I tried comparing `inode->m_filename` and `name`, although it would always return false, later I found out that this is caused by the fact that `KString` adds a null byte at the end of the string. The thing is that using `strncmp` works just fine so I'm thinking of leaving this as it is.

Reply:
That sounds like an issue with `KString` though (or rather, where that specific `KString` is constructed) if unexpected null bytes are appearing, as `KString` does store a length.